boot/bootutil/include/bootutil/sha256.h

```diff
@@ -31,14 +31,20 @@
 
 #include "mcuboot_config/mcuboot_config.h"
 
-#if (defined(MCUBOOT_USE_MBED_TLS) && defined(MCUBOOT_USE_TINYCRYPT))\
-    ||(defined MCUBOOT_USE_MBED_TLS) && defined(MCUBOOT_USE_CC310)\
-    ||(defined(MCUBOOT_USE_TINYCRYPT) && defined(MCUBOOT_USE_CC310))
-    #error "Cannot define CC310, MBED_TLS and TINYCRYPT"
+#if (defined(MCUBOOT_USE_MBED_TLS)    && defined(MCUBOOT_USE_TINYCRYPT))\
+    ||(defined MCUBOOT_USE_MBED_TLS)  && defined(MCUBOOT_USE_CC310)\
+    ||(defined(MCUBOOT_USE_TINYCRYPT) && defined(MCUBOOT_USE_CC310))\
+    ||(defined MCUBOOT_USE_MBED_TLS)  && defined(MCUBOOT_USE_SHARED_CRYPTO)\
+    ||(defined(MCUBOOT_USE_TINYCRYPT) && defined(MCUBOOT_USE_SHARED_CRYPTO))\
+    ||(defined(MCUBOOT_USE_CC310)     && defined(MCUBOOT_USE_SHARED_CRYPTO))
+    #error "Cannot define pairs of CC310, MBED_TLS, SHARED_CRYPTO and TINYCRYPT"
```

Comment:
`defined(x)` evaluates to 1 if x is defined, and 0 otherwise, so you could also write something like this to replace this block and the one below:

```
#define N_CRYPTO_BACKENDS (defined(MCUBOOT_USE_MBED_TLS) + \
			   defined(MCUBOOT_USE_TINYCRYPT) + \
			   defined(MCUBOOT_USE_CC310) + \
			   defined(MCUBOOT_USE_SHARED_CRYPTO))
#if N_CRYPTO_BACKENDS > 1
#error "Multiple crypto backends selected"
#elif N_CRYPTO_BACKENDS == 0
#error "You must select a crypto backend (CC310, MBED_TLS, SHARED_CRYPTO, and TINYCRYPT)"
#endif
```

This will grow linearly in the number of options, instead of quadratically like this strategy. Probably that's an upstream patch first, though.

Reply:
Yeah I know I've thought about it but I just wonder how many back-ends they actually want to support upstream. As far as I'm concerned they've seem to be pretty conservative to adding more back-end support. However since it'll make it easier to maintain the patch downstream easier I'll suggest the patch upstream and separate that into a separate [nrf fromlist] commit for this patch. 

I'll add the NRF prefix to it it's a good idea to separate out the namespace for what is NRF specific `MCUBOOT_USE_NRF_SHARED_CRYPTO` is that fine by you?

Reply:
> As far as I'm concerned they've seem to be pretty conservative to adding more back-end support. However since it'll make it easier to maintain the patch downstream easier I'll suggest the patch upstream and separate that into a separate [nrf fromlist] commit for this patch.

Sounds great, thank you. I'll add a comment on the upstream PR to see if those maintainers would be interested in abstracting this more.

> I'll add the NRF prefix to it it's a good idea to separate out the namespace for what is NRF specific MCUBOOT_USE_NRF_SHARED_CRYPTO is that fine by you?
 
Sure, and also `NRF_BOOT_USE_SHARED_CRYPTO` and `NRF_BOOT_SHARED_CRYPTO` within the Kconfig optinos. Thanks!

Reply:
Added upstream PR now here https://github.com/JuulLabs-OSS/mcuboot/pull/460 also appended [nrf fromlist] to the commit